File: VCSJones.FiddlerCertGen/X509AuthorityKeyIdentifierExtension.cs
using System;
using System.ComponentModel;
using System.Runtime.InteropServices;
using System.Security.Cryptography.X509Certificates;
using VCSJones.FiddlerCertGen.Interop;

namespace VCSJones.FiddlerCertGen
{
    public class X509AuthorityKeyIdentifierExtension : X509Extension
    {
        public X509AuthorityKeyIdentifierExtension(byte[] publicKeyId, byte[] serialNumber)
            : base(OIDs.szOID_AUTHORITY_KEY_IDENTIFIER2, EncodeData(publicKeyId, serialNumber), false)
        {
        }

        private static byte[] EncodeData(byte[] publicKey, byte[] serialNumber)
        {
            IntPtr publicKeyPtr = IntPtr.Zero, serialNumberPtr = IntPtr.Zero;
            try
            {
                var identifier = new CERT_AUTHORITY_KEY_ID2_INFO();
                identifier.AuthorityCertIssuer = new CERT_ALT_NAME_INFO { cAltEntry = 0 };
                if (serialNumber != null)
                {
                    serialNumberPtr = Marshal.AllocHGlobal(serialNumber.Length);
                    Marshal.Copy(serialNumber, 0, serialNumberPtr, serialNumber.Length);
                    identifier.AuthorityCertSerialNumber = new CRYPTOAPI_BLOB { cbData = (uint)serialNumber.Length, pbData = serialNumberPtr };

                }
                if (publicKey != null)
                {
                    publicKeyPtr = Marshal.AllocHGlobal(publicKey.Length);
                    Marshal.Copy(publicKey, 0, publicKeyPtr, publicKey.Length);
                    identifier.KeyId = new CRYPTOAPI_BLOB { cbData = (uint)publicKey.Length, pbData = publicKeyPtr };

                }
                uint dataSize = 0;
                IntPtr data;
                if (!Crypt32.CryptEncodeObjectEx(EncodingType.X509_ASN_ENCODING, OIDs.szOID_AUTHORITY_KEY_IDENTIFIER2, ref identifier, 0x8000, IntPtr.Zero, out data, ref dataSize))
                {
                    throw new Win32Exception(Marshal.GetLastWin32Error());
                }
                var buffer = new byte[dataSize];
                Marshal.Copy(data, buffer, 0, (int)dataSize);
                return buffer;
            }
            finally
            {
                Marshal.FreeHGlobal(publicKeyPtr);
                Marshal.FreeHGlobal(serialNumberPtr);
            }
        }
    }
}

File: VCSJones.FidderCertGenFS/X509AuthorityKeyIdentifierExtension.fs
namespace VCSJones.FiddlerCertGen

open System.Runtime.InteropServices
open System.Security.Cryptography
open System.Security.Cryptography.X509Certificates

module internal X509AuthorityKeyIdentifierEncoder = 
    let encodeExtensions(publicKey : byte array option, serialNumber : byte array option) =
        use serialNumberPtr = match serialNumber with
                                | None -> GlobalBufferSafeHandle.Null
                                | Some serial -> 
                                    let ptr = GlobalBufferSafeHandle.Allocate(serial.Length)
                                    Marshal.Copy(serial, 0, ptr.DangerousGetHandle(), serial.Length)
                                    ptr
        use publicKeyPtr = match publicKey with
                                | None -> GlobalBufferSafeHandle.Null
                                | Some key -> 
                                    let ptr = GlobalBufferSafeHandle.Allocate(key.Length)
                                    Marshal.Copy(key, 0, ptr.DangerousGetHandle(), key.Length)
                                    ptr
        let mutable identifier =
            CERT_AUTHORITY_KEY_ID2_INFO(
                CRYPTOAPI_BLOB(publicKey |?? (0u, fun arr -> uint32 arr.Length), publicKeyPtr.DangerousGetHandle()),
                CERT_ALT_NAME_INFO(cAltEntry = 0u),
                CRYPTOAPI_BLOB(serialNumber |?? (0u, fun arr -> uint32 arr.Length), serialNumberPtr.DangerousGetHandle())
            )
        let mutable buffer = LocalBufferSafeHandle.Null
        let mutable dataSize : uint32 = 0u
        if not <| ExternInterop.CryptEncodeObjectExAuthority(EncodingType.X509_ASN_ENCODING, OIDs.szOID_AUTHORITY_KEY_IDENTIFIER2, &identifier, 0x8000u, 0n, &buffer, &dataSize) then
            invalidOp("Failed to encode object.")
        using(buffer) (fun _ ->
            let dataBuffer = Array.zeroCreate<byte>(int dataSize)
            Marshal.Copy(buffer.DangerousGetHandle(), dataBuffer, 0, int dataSize)
            dataBuffer
        )

type X509AuthorityKeyIdentifierExtension(publicKey : byte array option, serialNumber : byte array option) = 
    inherit X509Extension(new Oid(OIDs.szOID_AUTHORITY_KEY_IDENTIFIER2), X509AuthorityKeyIdentifierEncoder.encodeExtensions(publicKey, serialNumber), false)